platforms/software/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/DefaultCompositeTestEventGenerator.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.testing;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.api.tasks.testing.CompositeTestEventGenerator;
+import org.gradle.api.tasks.testing.TestEventGenerator;
+import org.gradle.api.tasks.testing.TestFailure;
+import org.gradle.api.tasks.testing.TestOutputEvent;
+import org.gradle.api.tasks.testing.TestResult;
+import org.gradle.internal.id.IdGenerator;
+
+import javax.annotation.Nullable;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+@NonNullApi
+public final class DefaultCompositeTestEventGenerator implements CompositeTestEventGenerator {
```

Comment:
I think we want 3 implementations of this.

- DefaultTestEventGenerator that implements started/output/terminal methods. This enforces the lifecycle for each of those methods. close() fails if started or a terminal method has never been called.
- DefaultCompositeTestEventGenerator. everything as above, except it allows you to create children as long as the composite has started and has not reached a terminal method. close() closes all children
- DefaultRootTestEventGenerator.  everything as above, except close() checks if there were any failures and throws a VerificationException if so